Added waterfall display to webwenet WEBUI

Tysonpower opened the pull request:
I really like the web decoder and had a go at a local balloon recently getting my first wenet images received. I found it a bit hard to aim my yagi by hand without a waterfall, so i added one to the webui :)

Changes:
* Added waterfall like heatmap matrix to webui
* expanded area for spectrum from 150px to 300px to fit both
* changed fft update from 1000ms to 500ms

Can be tested here: https://wenet.do5ty.de/

<img width="1794" height="701" alt="image" src="https://github.com/user-attachments/assets/eb8d098a-c079-45e2-80cc-1a9db6df4e38" />

Tysonpower committed:
Added waterfall display

darksidelemm commented:
Here's what it looks like when testing with a Wenet v2 signal:

<img width="1483" height="625" alt="Screen Shot 2025-10-01 at 07 47 21" src="https://github.com/user-attachments/assets/ccf10783-03db-4708-9dfc-9deba14dc92a" />

It's pretty cool!

I'd be somewhat inclined to drop the spectrum display altogether in this case, since otherwise a fair bit of screen real estate is used up.

darksidelemm commented:
Have you done any comparisons of CPU usage between this and without the waterfall?

TheSkorm commented:
This is nice and certainly worth adding in. I don't have time immediately to look into the PR but have added it into my todo list so it doesn't get lost.


Things I think we should do/check/change
- [ ] Check that the CPU usage is comparable (use web dev tools for this). I had a spectrum plot using canvas in a much older version (https://github.com/projecthorus/webhorus/commit/704075c19875363011ae78d7a43e3b2c26ac030b) but wasn't happy with it.
- [ ] Drop the spectrum display as it would be unnecessary now
- [ ] The frequency estimator marks should either be moved to little markers at the edge of the waterfall or turned into a line that's tracked with history so that it makes sense on the waterfall
- [ ] Make sure this works with horus in both rtl-sdr and audio modes (sorry the code base is a mess, wenet and rtl-sdr support is something we never thought would be possible...)
- [ ] We might want to tweak the update rate depending on cpu usage to provide a smoother plot?
- [ ] Confirm that resizing the window is working ok, plotly is a little annoying with that.

I'm happy to work on this, but might take me a little while to get to

Tysonpower commented:
Only using the waterfall with slow updates as before it uses only a bit more CPU, 19% of time instead of 13%.

Update rate at 500ms will double that of course - i think 250ms update rate would be even nicer - maybe we should add a setting for that so users can use the fast update rate if their system is fast enough.

Replacing the spectrum view graph with the waterfall is a good idea, will work on that, maybe even putting it into all modes including Horus.

Tysonpower committed:
Removed old spectrum, add fft interval setting

Tysonpower committed:
Added waterfall to horus SDR and Audio

Tysonpower commented:
I think i got all the points mentioned implemented now, updaterate in horus is the same, cpu usage as well after removing the old spectrum display.

In wenet the default updaterate is 500ms now and it can be changed in the config settings.

Testing would be appreciated, still need to get my wenet TX setup.

> Things I think we should do/check/change
> 
> * [X]  Check that the CPU usage is comparable (use web dev tools for this). I had a spectrum plot using canvas in a much older version ([704075c](https://github.com/projecthorus/webhorus/commit/704075c19875363011ae78d7a43e3b2c26ac030b)) but wasn't happy with it.
> * [X]  Drop the spectrum display as it would be unnecessary now
> * [X]  The frequency estimator marks should either be moved to little markers at the edge of the waterfall or turned into a line that's tracked with history so that it makes sense on the waterfall
> * [X]  Make sure this works with horus in both rtl-sdr and audio modes (sorry the code base is a mess, wenet and rtl-sdr support is something we never thought would be possible...)
> * [X]  We might want to tweak the update rate depending on cpu usage to provide a smoother plot?
> * [X]  Confirm that resizing the window is working ok, plotly is a little annoying with that.
>

darksidelemm commented:
Visually this is looking pretty cool. I've had it running in both audio and RTLSDR modes.
A few comments:
- The waterfall display kind of fills up and 'squishes in' on startup, until ~120s of data is reached. Ideally this should just just scroll down from the top instead of being expanded to fill the area.
- The frequency estimator probably shouldn't be shown as vertical lines over the waterfall. It should just be points (down arrows?) at the top of the waterfall.
- We've also now lost any indication of where the frequency estimator bounds are, which is important in Horus Binary mode to indicate where the signal needs to lie within the passband to be decoded. So some new way of indicating this range is required.
- I'm not really a fan of the Portland colormap. I'd prefer either Jet or Turbo (which I know isn't available by default within plotly).

Tysonpower commented:
Thanks for the feedback, will work on that tomorrow - also just got my first wenetV2 signal send across my desk :)

Tysonpower committed:
Changed color to Turbo, added arrows, fixed bugs

Tysonpower commented:
Fixed all the points i think, updated my site as well - here is a screenshot receiving test images:

<img width="1640" height="730" alt="image" src="https://github.com/user-attachments/assets/ad839f3a-d611-4bbf-b1cd-0e374e692f96" />

TheSkorm commented:
Made some tweaks to hopefully improve performance a bit. not sure how it'll go. have push a build up.

Tysonpower commented:
Thanks for accepting the PR :)
